downloader/github.go

```diff
@@ -0,0 +1,195 @@
+		"GithubRepo": cfg.GithubRepo,
+	}
+
+	var missingConfigs []string
+
+	// Check for configs that are not set
+	for cfgName, cfgValue := range requiredConfigs {
+		if cfgValue == "" {
+			missingConfigs = append(missingConfigs, cfgName)
+		}
+	}
+
+	// Error on all the missing configs
+	if len(missingConfigs) > 0 {
+		return fmt.Errorf("configuration values cannot be empty: %s", strings.Join(missingConfigs, ", "))
+	}
+
+	return nil
+}
+
+// createGithubClient creates a new Github client to be used by the github downloader
+func createGithubClient(accessToken string) *github.Client {
+	ctx := context.Background()
+	if accessToken != "" {
+		ts := oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: accessToken},
+		)
+		tc := oauth2.NewClient(ctx, ts)
+		return github.NewClient(tc)
+	}
+
+	return github.NewClient(nil)
+}
+
+// buildObjectPath builds a path to an object that matches the Artifactory path
+func (ghd *githubDownloader) buildObjectPath(tagName string, assetName string) string {
+	return fmt.Sprintf("%s/%s/%s/%s", ghd.repoOwner, ghd.repoName, tagName, assetName)
+}
+
+// ListObjects lists the objects available in the Github Repo
+func (ghd *githubDownloader) ListObjects() ([]string, error) {
+	var objects []string
+
+	ctx := context.Background()
+
+	releases, _, err := ghd.client.Repositories.ListReleases(ctx, ghd.repoOwner, ghd.repoName, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, release := range releases {
+		assets, _, err := ghd.client.Repositories.ListReleaseAssets(ctx, ghd.repoOwner, ghd.repoName, *release.ID, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, asset := range assets {
+			objects = append(objects, ghd.buildObjectPath(*release.TagName, *asset.Name))
+		}
+	}
+	return objects, nil
+}
+
+// getReleaseID Gets the ID of a release from the release tag
+func (ghd *githubDownloader) getReleaseID(releaseTag string) (int64, error) {
+	ctx := context.Background()
+
+	releases, _, err := ghd.client.Repositories.ListReleases(ctx, ghd.repoOwner, ghd.repoName, nil)
+	if err != nil {
+		return -1, err
+	}
+
+	for _, release := range releases {
+		if *release.TagName == releaseTag {
+			return *release.ID, nil
+		}
+	}
+	return -1, fmt.Errorf("release '%s' not found", releaseTag)
+}
+
+// getAssetID Gets the ID of a asset from the provided assetName for the given releaseID
+func (ghd *githubDownloader) getAssetID(releaseID int64, assetName string) (int64, error) {
+	ctx := context.Background()
+
+	assets, _, err := ghd.client.Repositories.ListReleaseAssets(ctx, ghd.repoOwner, ghd.repoName, releaseID, nil)
+	if err != nil {
+		return -1, err
+	}
+
+	for _, asset := range assets {
+		if *asset.Name == assetName {
+			return *asset.ID, nil
+		}
+	}
+	return -1, fmt.Errorf("asset '%s' not found", assetName)
+}
+
+// GetObject downloads the object specified in sourceObj to the targetPath
+func (ghd *githubDownloader) GetObject(sourceObj string, targetPath string) error {
```

Comment:
I think if we create a 3rd downloader, we should evaluate if we want to create an interface called Downloader or something like that where the downloaders can have a generalized interface to implement, and we can factor out some shared code. That could also make it a lot easier to implement new downloaders.

For now, probably wait and see what ends up being common between everything.